src/inmanta/references.py

```diff
@@ -0,0 +1,401 @@
+"""
+    Copyright 2024 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+import abc
+import collections
+import inspect
+import typing
+import uuid
+
+import pydantic
+
+import inmanta
+from inmanta.util import dict_path
+
+# from inmanta.dataclass import Value  # The base class for data classes: see the specific design
+
+ReferenceType = typing.Annotated[str, pydantic.StringConstraints(pattern="^([a-z0-9_]+::)+[A-Z][A-z0-9_-]*$")]
+PrimitiveTypes = str | float | int | bool
+
+
+class Argument(pydantic.BaseModel):
+    """Base class for reference (resolver) arguments"""
+
+    type: str
+    """ The type of the argument. It is used a discriminator to select the correct class
+    """
+
+    name: str
+    """ The name of the argument in the reference constructor
+    """
+
+    # TODO: we probably need a context like object so that it is easier to pass more data to this method
+    @abc.abstractmethod
+    def get_arg_value(self, resource: "resources.Resource", references: dict[uuid.UUID, "Reference"]) -> object:
+        """Get the value for the argument to be able to construct the reference again
+
+        :param resource: The resource on which the reference has been defined
+        :param references: A mapping from id to reference. This mapping can be empty when this method is called. This mapping
+                           is only complete when the full deserialization process is complete.
+        """
+
+
+class LiteralArgument(Argument):
+    """Literal argument to a reference"""
+
+    type: typing.Literal["literal"] = "literal"
+    value: PrimitiveTypes
+
+    def get_arg_value(self, resource: "resources.Resource", references: dict[uuid.UUID, "Reference"]) -> object:
+        return self.value
+
+
+class ReferenceArgument(Argument):
+    """Use the value of another reference as an argument for the reference"""
+
+    type: typing.Literal["reference"] = "reference"
+    id: uuid.UUID
+
+    def get_arg_value(self, resource: "resources.Resource", references: dict[uuid.UUID, "Reference"]) -> object:
+        return ReferencePlaceHolder(self.id, references)
```

Comment:
I would expect this to return the actual value (not a placeholder)

Reply:
The problem is that we do not know if the reference is already deserialized.